### PR TITLE
[FLINK-15910] [legal] Add NOTICE and LICENSE files

### PR DESCRIPTION
**Pull request by tzulitai**
The Apache Parent POM added in #12 already pulls in default LICENSE / NOTICE files for built Maven artifact jars.

This PR explicitly adds NOTICE / LICENSE files for the following distribution purposes:

**Canoncial source distribution**

0659c0d Adds NOTICE file to be shipped with the source distribution.
The only dependency we bundle in the source is copied & modified source code from the fastutils project (http://fastutil.di.unimi.it/), so this is mentioned.
See the copied code under `statefun-flink/statefun-flink-core`.

**Maven artifacts**

9161d70
The only artifact jar that bundles dependencies is `statefun-flink-distribution`, so an explicit NOTICE file is added for it under `META-INF`.
Additionally, since some BSD licensed dependencies are bundled, their respective licenses can also be found under `META-INF/licenses` to be bundled with the jar.

### CODE REVIEW

**Comment by tzulitai** [comment]
cc @igalshilman

Also cc @zentol, if you find the time, would be great to have your input here to see if we're missing anything else legal wise.

File: statefun-flink/statefun-flink-distribution/src/main/resources/META-INF/NOTICE [hunk 1]
```diff
@@ -0,0 +1,40 @@
+statefun-flink-distribution
+Copyright 2014-2020 The Apache Software Foundation
+
+This product includes software developed at
+The Apache Software Foundation (http://www.apache.org/).
+
+This project bundles the following dependencies from (or other subprojects of) the owning top-level project, Apache Flink,
+under the Apache Software License 2.0 (http://www.apache.org/licenses/LICENSE-2.0.txt)
```
Comment (tzulitai) [inline statefun-flink/statefun-flink-distribution/src/main/resources/META-INF/NOTICE]: @zentol do you think we actually need this whole section regarding bundled `flink` / `flink-shaded` dependencies?

File: statefun-flink/statefun-flink-distribution/src/main/resources/META-INF/NOTICE [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+- org.apache.flink:flink-annotations:1.10-SNAPSHOT
```
Comment (tzulitai) [inline statefun-flink/statefun-flink-distribution/src/main/resources/META-INF/NOTICE]: `1.10-SNAPSHOT` should be updated once Flink 1.10 is released and we no longer rely on the snapshots.

File: NOTICE [hunk 3]
```diff
@@ -0,0 +1,9 @@
+Apache Flink
+Copyright 2014-2020 The Apache Software Foundation
+
+This product includes software developed at
+The Apache Software Foundation (http://www.apache.org/).
+
+This project contains copied and modified source code that is licensed under the Apache Software License 2.0 (http://www.apache.org/licenses/LICENSE-2.0.txt) from the following projects:
+
+- fastutil (http://fastutil.di.unimi.it/) Copyright (c) - Sebastiano Vigna
```
Comment (zentol) [inline NOTICE]: Are the files in question tagged appropriately? For maintenance purposes you always want to be able to trace down the files from the NOTICE.

File: pom.xml [hunk 4]
```diff
@@ -187,6 +187,8 @@ under the License.
                         <exclude>statefun-docs/docs/_templates/**</exclude>
                         <!-- Generated code  -->
                         <exclude>**/generated/**</exclude>
+                        <!-- Bundled license files -->
+                        <exclude>**/LICENSE*</exclude>
```
Comment (zentol) [inline pom.xml]: exclude source NOTICE as well?

File: statefun-flink/statefun-flink-distribution/src/main/resources/META-INF/NOTICE (hunk 1 above)
Comment (zentol) [reply statefun-flink/statefun-flink-distribution/src/main/resources/META-INF/NOTICE]: If the shade-plugin is configured correctly then the Flink entries will be added automatically upon inclusion.

You want to add this transformer:
```
<transformer implementation="org.apache.maven.plugins.shade.resource.ApacheNoticeResourceTransformer">
	<projectName>Apache Flink-statefun</projectName>
	<encoding>UTF-8</encoding>
</transformer>
```

File: NOTICE (hunk 3 above)
Comment (tzulitai) [reply NOTICE]: @zentol what kind of tagging do you have in mind?
Would a mention in NOTICE of where the copied files are in the source be sufficient?

File: pom.xml (hunk 4 above)
Comment (tzulitai) [reply pom.xml]: @zentol It seems like NOTICE / LICENSE files without any suffixes are automatically ignored by the rat plugin already.

File: NOTICE (hunk 3 above)
Comment (zentol) [reply NOTICE]: A small note in the javadocs + paths in the notice, to cover all our bases.

Comment (tzulitai) [reply NOTICE]: Got it, thanks

**Comment by tzulitai** [comment]
Thanks for the review @zentol!
All comments have been addressed. I'll proceed to merge this once Travis is green.

**Pull request closed without merging**